### ISSUE TITLE
Spurious attempt to load a crate resulting in "multiple rlib candidates for `debug_unreachable` found".

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#56590]
I've reduced a triggering example to this:
```rust
use a::A;
type B<'a, T> = C<&'a T>;
```
Producing ([on playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=67d1e5afa43f3b28e938d58338479015), in Rust 2018 mode, only on nightly):
```
error[E0465]: multiple rlib candidates for `debug_unreachable` found
  |
note: candidate #1: /playground/target/debug/deps/libdebug_unreachable-dc2ba43f474707f3.rlib
note: candidate #2: /playground/target/debug/deps/libdebug_unreachable-fedb7fe62f71c3f0.rlib
```

The names in the testcase appear to not matter, but I can't figure out why both a lifetime and a type parameter need to be present, or any of the other details.

My guess this is from the `rustc_resolve` speculative loading of crates, to give better suggestions, which I must've not made foolproof (I *thought* I made all the errors lazy, but maybe this from a different spot that I missed).

cc @petrochenkov

**Comment by petrochenkov** [comment]
Crate loading is certainly not fully speculative, at leas with regards to non-existent paths (https://github.com/rust-lang/rust/issues/55103), so it's possible that some other errors are also reported eagerly.

**Comment by Arnavion** [comment]
This appears to be fixed in nightly, but the original issue is now in beta.

**Comment by shepmaster** [comment]
And stable 1.31.1

**Comment by xftroxgpx** [comment]
_nightly release_ still has this, but _nightly debug_ does not! (I hope it's not because I ran debug first and release later)
(also _stable debug_ and _beta debug_ still have this, but it might be because I ran their release before their debug? so did release create one rlib and debug the second? is that why the error is on the debug then?)

I hit it with my example too ([playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=0f76b7f6813ab74426748c88d5a8775b)):
```rust

#[derive(Debug)]
struct A(i32);

fn main() {
    let mut v = vec![A(1), A(2)];

    //let first = &v[0];

    v.push(A(6));
    //let first = &v[0]; //all errors go away if I uncomment this, even the "multiple rlib" one
    

    println!("The first element is: {:?}", first);
}
```

**Comment by xftroxgpx** [comment]
Question: Does [this](https://github.com/rust-lang/rust/issues/39279#issuecomment-275545608) explain it? ie. is it actually a [cargo bug](https://github.com/rust-lang/cargo/issues/3599)?

**Comment by jethrogb** [comment]
This is now on stable, but it appears fixed on nightly

**Comment by shepmaster** [comment]
> but it appears fixed on nightly. 


[My code example](https://github.com/rust-lang/rust/issues/56590#issuecomment-473536351) continues to have the error in nightly on the playground.

**Comment by lqd** [comment]
The code example in https://github.com/rust-lang/rust/issues/56590#issuecomment-450535106 still fails on nightly as well, but in Release mode only, while some examples from other comments fail in Debug only, and finally there are also examples (such as Shep's) failing in both Debug and Release.

**Comment by garrettmaring** [comment]
Is there a known workaround for this? I am still seeing this issue with 1.44.0-nightly.

**Comment by petrochenkov** [comment]
Fixed in https://github.com/rust-lang/rust/pull/74071.
